### ISSUE TITLE
after docker daemon restart, docker system prune -f sometimes deletes used volumes

### ISSUE THREAD

**Issue by ffdybuster** [moby/moby#38883]
**Description**

If I restart the docker daemon and later do a `docker system prune -f`, volumes of containers which have already been running before the restart are removed.

**Steps to reproduce the issue:**
1. start `redis:latest` with no volumes (i.e. data will be written to anonymous volume)
2. `sudo systemctl restart docker`
3. `docker system prune -f`

**Describe the results you received:**

The anonymous volume used by the container is gone.

**Describe the results you expected:**

The anonymous volume used by the container is still there (since the container is up and running).

**Additional information you deem important (e.g. issue happens only occasionally):**

I've only experienced this since the last update IIRC. I first thought it was a random glitch, but then it happened again after I restarted the daemon on several machines yesterday and a nightly cronjob did the pruning.

**Output of `docker version`:**

```
Client:
 Version:           18.09.3
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        774a1f4
 Built:             Thu Feb 28 06:34:04 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.3
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       774a1f4
  Built:            Thu Feb 28 05:59:55 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 26
 Running: 23
 Paused: 0
 Stopped: 3
Images: 167
Server Version: 18.09.3
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: e6b3f5632f50dbc4e9cb6288d911bf4f5e95b18e
runc version: 6635b4f0c6af3810594d2770f662f34ddc15b40d
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.0-8-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 62.88GiB
Name: xxxx
ID: yyyy
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Registry Mirrors:
 http://127.0.0.1:5004/
Live Restore Enabled: true
Product License: Community Engine

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Physical server, no virtualization (except docker running on it).

**Comment by fhackenberger** [comment]
This bug just hit me, I lost 19GB of production email data!!

**Comment by denji** [comment]
rm_rf "by Design".

**Comment by fhackenberger** [comment]
Just to clarify, the `docker volume prune` command says:
`# docker volume prune
WARNING! This will remove all local volumes not used by at least one container.
Are you sure you want to continue? [y/N] y`
and the volumes I lost were definitely used by running containers. I was literally using the mailcow interface, while the underlying DB data was removed.

**Comment by shachibista** [comment]
How about list the volumes to be deleted instead of a simple prompt?

**Comment by felixfontein** [comment]
That won't work so simply; the way prune is implemented, the `docker` CLI command tells the docker daemon to do the pruning, and gets returned a list of containers / volumes / ... that were pruned. There's no dry-run version of that (yet), so either the CLI has to re-implement pruning to get hold of this list, or the daemon has to be modified.

Anyway, the problem here is that the daemon deletes volumes which are still in use, despite API documentation saying that it should not. (https://docs.docker.com/engine/api/v1.39/#operation/VolumePrune)

**Comment by alexandrestein** [comment]
I lost lots of data.
This is a critical bug.

Multiple running container has lost there volume with all the data it was containing...

**Comment by bsousaa** [comment]
Can you confirm if this issue is reproducible in a current version of Docker?

**Comment by ffdybuster** [comment]
I currently cannot run the latest version due to #44970. I can try again once the next bugfix release is out including a fix for that.

**Comment by sam-thibault** [comment]
The #44970 fix was made in 23.0.2 if you'd like to try again.

**Comment by ffdybuster** [comment]
I wasn't able to reproduce it anymore with 23.0.3 (I also additionally run `docker volume prune`, the volume's still there). Therefore closing...
